keps/sig-apps/20190305-add-progress-deadline-seconds-to-job.md

```diff
@@ -0,0 +1,131 @@
+For example, we create a job with miss-typed pull spec, and will find out about it only when the job controller kicks a pod and the pod will result in ImagePullBackOff state. The job will remain active.
+Users would like to fail the job instead of having the job remain active if there's a typo, config error, not-pullable image etc.
+For detailed examples, please see the below user stories.
+
+[#48075]: https://github.com/kubernetes/kubernetes/issues/48075
+[#51153]: https://github.com/kubernetes/kubernetes/issues/51153
+
+### Goals
+
+We'd like to fail a job if it does not start within given time.
+
+### Non-Goals
+
+N/A
+
+## Proposal
+
+### User Stories
+
+* [#67828][] `Jobs with a container config error will never complete.` A job with a config error (pod: Warning Failed 22m (x8 over 24m) kubelet, minikube Error: container has runAsNonRoot and image will run as root), will never complete, the job will not have any success or fail status, and will have an 'activate' forever. While this should count as a failure, and take into account restartPolicy and backoffLimit.
+* [#62816][] `Request for job failure on pod failure rather than job remaining active.` 
+I run a job with a volume that references a non-existent config map and get an error message like "MountVolume.SetUp failed for volume "some-volume" : configmaps "some-missing-config-map" not found". I expect the status returned by an API call to have "failed": 1.
+Would also like the job to fail if the container image can't be pulled. Would be nice to be able to specify in the job spec how many image pull attempts to make before failing. If I have a typo in the image name, or it's not accessible in my Docker registry, I want the job to fail so I can find that out immediately at job submission/creation time, without having to occasionally look through the job list in the dashboard to find ones with the failures recorded in their details.
+
+[#67828]: https://github.com/kubernetes/kubernetes/issues/67828
+[#62816]: https://github.com/kubernetes/kubernetes/issues/62816
+
+### Detailed Design 
+
+We will add the following API fields to `JobSpec` (`Job`'s `.spec`).
+
+```go
+type JobSpec struct {
+     // Optional duration in seconds relative to the startTime that the job needs to become active,
+     // otherwise the system will try to terminate it; value must be positive integer
+     // Job is `active` means all pods of the job are in `running` state.
+     // +optional
+     ProgressDeadlineSeconds *int64
+}
+```
+
+Job is `active` means all pods of the job are in `running` state.
+
+### Implementation Details/Notes/Constraints
+
+#### Implementation Details
+We'll check the job status, and if the following three criterions are met, the job is considered failure.
+*  job has not retried, i.e, the job is on its first run
+*  job has pending pods, i.e, not all pods are running
+*  job duration has exceeded `*job.Spec.ProgressDeadlineSeconds`
```

Comment:
Job duration might be confusing with `ActiveDeadlineSeconds`, clearly state this is the time from job's startTime until the moment the check if performed.